assignment/core/src/main/java/com/assignment/core/services/CipherServiceImpl.java

```diff
@@ -0,0 +1,59 @@
+package com.assignment.core.services;
+
+
+import com.assignment.core.crypto.AsymmetricCipherProvider;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Component(service = CipherService.class)
+public class CipherServiceImpl implements CipherService {
+
+    @Reference(
+            service = AsymmetricCipherProvider.class,
+            cardinality = ReferenceCardinality.MULTIPLE,
+            policy = ReferencePolicy.DYNAMIC
+    )
+    private volatile Map<String, AsymmetricCipherProvider> cipherProviders = new ConcurrentHashMap<>();
```

Comment:
Why not simply inject them into a List, as discussed in the training session ?

Reply:
injected them into list  please review changes